### PR TITLE
[6.8.z] pytest conversion - api.test_discoveredhost

### PR DESCRIPTION
**Pull request by rplevka**
backport of
https://github.com/SatelliteQE/robottelo/pull/8112

results

### CODE REVIEW

File: tests/foreman/api/test_discoveredhost.py [hunk 1]
```diff
@@ -68,561 +76,549 @@ def _create_discovered_host(name=None, ipaddress=None, macaddress=None):
...
+            raise AssertionError(f'Timed out waiting for {pattern[1]} from VM')
+    groups = re.search('DHCPACK on (\\d.+) to', dhcp_fdi.out)
+    assert len(groups.groups()) == 1, 'Unable to parse FDI ip address'
+    fdi_ip = groups.groups()[0]
+
+    # finally, assert that the FDI successfully uploaded its facts to the server
+    try:
+        facts_fdi = _wait_for_log(
+            channel,
+            f'\\[I\\|app\\|[a-z0-9]+\\] Started POST '
```
Comment (mshriver) [inline tests/foreman/api/test_discoveredhost.py:169]: This `f` prefix isn't necessary but is completely inconsequential.

**Review by mshriver — APPROVED** [review]

File: tests/foreman/api/test_discoveredhost.py [hunk 2]
```diff
@@ -68,561 +76,549 @@ def _create_discovered_host(name=None, ipaddress=None, macaddress=None):
...
+    # read lines until the buffer is empty
+    while ch.recv_ready():
+        log_line = ''
+        # read bytes one-by-one until we have a complete log line
+        while ch.recv_ready():
+            char_encoded = ch.recv(1)
+            try:
+                char = char_encoded.decode('utf-8')
+            except UnicodeEncodeError:
+                logger.warning(f'Failed to decode char using utf-8 charset, replacing by "?"')
```
Comment (JacobCallahan) [inline tests/foreman/api/test_discoveredhost.py:90]: no f-string needed here, unless you were wanting to add in the offending line.

File: tests/foreman/api/test_discoveredhost.py [hunk 3]
```diff
@@ -68,561 +76,549 @@ def _create_discovered_host(name=None, ipaddress=None, macaddress=None):
...
+        try:
+            _wait_for_log(channel, pattern[0], timeout=120)
+        except TimedOutError:
+            # raise assertion error
+            raise AssertionError(f'Timed out waiting for VM (tftp) to fetch {pattern[1]}')
+
+    # assert that server receives DHCP discover from FDI
+    for pattern in [
+        (f"DHCPDISCOVER from {host.mac}", "DHCPDISCOVER",),
+        (f"DHCPACK on [0-9.]+ to {host.mac}", "DHCPACK"),
```
Comment (JacobCallahan) [inline tests/foreman/api/test_discoveredhost.py:154]: single quotes, please

File: tests/foreman/api/test_discoveredhost.py [hunk 4]
```diff
@@ -68,561 +76,549 @@ def _create_discovered_host(name=None, ipaddress=None, macaddress=None):
...
+    3. Update Global parameters to set default org and location for
+       discovered hosts.
+    4. Enable auto_provision flag to perform discovery via discovery
+       rules.
+    """
+    # Build PXE default template to get default PXE file
+    entities.ProvisioningTemplate().build_pxe_default()
+    # let's just modify the timeouts to speed things up
+    ssh.command(
+        "sed -ie 's/TIMEOUT [[:digit:]]\\+/TIMEOUT 1/g' " "/var/lib/tftpboot/pxelinux.cfg/default"
```
Comment (JacobCallahan) [inline tests/foreman/api/test_discoveredhost.py:240]:
```suggestion
        "sed -ie 's/TIMEOUT [[:digit:]]\\+/TIMEOUT 1/g' /var/lib/tftpboot/pxelinux.cfg/default"
```

File: tests/foreman/api/test_discoveredhost.py [hunk 5]
```diff
@@ -68,561 +76,549 @@ def _create_discovered_host(name=None, ipaddress=None, macaddress=None):
...
-            be discovered via PXE boot.
+    :CaseImportance: High
+    """
+    name = gen_choice(list(valid_data_list().values()))
+    result = _create_discovered_host(name)
 
-        :Steps: PUT /api/v2/discovered_hosts/:id/reboot
+    entities.DiscoveredHost(id=result['id']).delete()
+    search = entities.DiscoveredHost().search(
+        query={'search': 'name == {}'.format(result['name'])}
```
Comment (JacobCallahan) [inline tests/foreman/api/test_discoveredhost.py:410]: can do an f-string here

File: tests/foreman/api/test_discoveredhost.py [hunk 6]
```diff
@@ -68,561 +76,549 @@ def _create_discovered_host(name=None, ipaddress=None, macaddress=None):
...
+            discovered_host = _assert_discovered_host(pxe_host, channel, user_config=cfg)
+            discovered_host.reboot()
 
-        :expectedresults: User should be able to update the rule and it should
-            be applied on discovered host
+            # assert that server receives DHCP discover from hosts PXELinux
+            # this means that the host got rebooted
+            for pattern in [
+                (f"DHCPDISCOVER from {pxe_host.mac}", "DHCPDISCOVER",),
+                (f"DHCPACK on [0-9.]+ to {pxe_host.mac}", "DHCPACK"),
```
Comment (JacobCallahan) [inline tests/foreman/api/test_discoveredhost.py:560]: single quotes here

File: tests/foreman/api/test_discoveredhost.py [hunk 7]
```diff
@@ -68,561 +76,549 @@ def _create_discovered_host(name=None, ipaddress=None, macaddress=None):
...
+                if is_open('BZ:1893349'):
+                    pass
+                else:
+                    raise e
+            # assert that server receives DHCP discover from hosts PXELinux
+            # this means that the hosts got rebooted
+            for pxe_host in [(pxe_host_1, channel_1), (pxe_host_2, channel_2)]:
+                for pattern in [
+                    (f"DHCPDISCOVER from {pxe_host[0].mac}", "DHCPDISCOVER",),
+                    (f"DHCPACK on [0-9.]+ to {pxe_host[0].mac}", "DHCPACK"),
```
Comment (JacobCallahan) [inline tests/foreman/api/test_discoveredhost.py:616]: single quotes here

**Review by JacobCallahan — CHANGES_REQUESTED** [review]

**Comment by rplevka** [comment]
I believe the travis failure is unrelated

**Comment by omkarkhatavkar** [comment]
Can one of the admins verify this patch?

File: robottelo/libvirt_discovery.py [hunk 8]
```diff
@@ -132,8 +141,8 @@ def create(self):
 
         if self.extra_nic:
             nic_mac = _gen_mac_for_libvirt()
-            command_args.append('--network=bridge:{vm_bridge}')
-            command_args.append('--mac={0}'.format(nic_mac))
+            command_args.append('--network={vm_nw_type}:{vm_network}')
```
Comment (mshriver) [inline robottelo/libvirt_discovery.py:144]: Missing an f-string here.

File: robottelo/libvirt_discovery.py [hunk 9]
```diff
@@ -181,14 +191,17 @@ def attach_nic(self):
         command_args = [
             'virsh attach-interface',
             '--domain={vm_name}',
-            '--type=bridge',
-            '--source={vm_bridge}',
+            '--type={vm_nw_type}',
```
Comment (mshriver) [inline robottelo/libvirt_discovery.py:194]: missing f-strings for these variables.

File: tests/foreman/api/test_discoveredhost.py [hunk 10]
```diff
@@ -68,561 +76,547 @@ def _create_discovered_host(name=None, ipaddress=None, macaddress=None):
...
+    for pattern in [
+        (f'DHCPDISCOVER from {host.mac}', 'DHCPDISCOVER',),
+        (f'DHCPACK on [0-9.]+ to {host.mac}', 'DHCPACK'),
+    ]:
+        try:
+            dhcp_fdi = _wait_for_log(channel, pattern[0], timeout=30)
+        except TimedOutError:
+            # raise assertion error
+            raise AssertionError(f'Timed out waiting for {pattern[1]} from VM')
+    groups = re.search('DHCPACK on (\\d.+) to', dhcp_fdi.out)
```
Comment (mshriver) [inline tests/foreman/api/test_discoveredhost.py:161]: I'd recommend using `r''` string prefix here to avoid warnings on escape characters.

Comment (mshriver) [reply tests/foreman/api/test_discoveredhost.py:161]: Applies to all regex expressions in this PR.

File: tests/foreman/api/test_discoveredhost.py [hunk 11]
```diff
@@ -68,561 +76,547 @@ def _create_discovered_host(name=None, ipaddress=None, macaddress=None):
...
+    return discovered_host[0]
+
+
+@pytest.fixture(scope="module", params=['non-admin', 'admin'])
+def _module_user(request, module_org, module_location):
+    if request.param == 'admin':
+        yield None
+    else:
+        # setup a clone of DiscoveryManager role
+        discoman_role = entities.Role().search(query={'search': 'name="Discovery Manager"'})[0]
```
Comment (mshriver) [inline tests/foreman/api/test_discoveredhost.py:214]: `discoman`, my favorite super-hero.

File: tests/foreman/api/test_discoveredhost.py [hunk 12]
```diff
@@ -68,561 +76,547 @@ def _create_discovered_host(name=None, ipaddress=None, macaddress=None):
...
+    host_name = 'mac{}'.format(discovered_host['mac'].replace(':', ''))
+    assert discovered_host['name'] == host_name
 
-    @pytest.mark.stubbed
-    @tier3
-    def test_positive_provision_pxe_less_host(self):
-        """Provision a pxe-less discovered hosts
 
-        :id: 91bb254b-3c30-4608-b1ba-e18bcc22efb5
+@pytest.mark.stubbed
```
Comment (mshriver) [inline tests/foreman/api/test_discoveredhost.py:315]: Duplicate markers

**Review by mshriver — CHANGES_REQUESTED** [review]
Couple f-strings to resolve, backport looks good otherwise. I think most/all of @JacobCallahan comments have been addressed.

File: tests/foreman/api/test_discoveredhost.py [hunk 13]
```diff
@@ -68,561 +76,547 @@ def _create_discovered_host(name=None, ipaddress=None, macaddress=None):
...
-                result = _create_discovered_host(name)
-                discovered_host = entities.DiscoveredHost(id=result['id']).read_json()
-                host_name = 'mac{0}'.format(discovered_host['mac'].replace(':', ''))
-                self.assertEqual(discovered_host['name'], host_name)
+    :BZ: 1731112
+    """
+    name = gen_choice(list(valid_data_list().values()))
+    result = _create_discovered_host(name)
+    discovered_host = entities.DiscoveredHost(id=result['id']).read_json()
+    host_name = 'mac{}'.format(discovered_host['mac'].replace(':', ''))
```
Comment (JacobCallahan) [inline tests/foreman/api/test_discoveredhost.py:311]:
```suggestion
    host_name = f'mac{discovered_host["mac"].replace(":", "")}'
```

**Pull request closed without merging**